File: python/maze.py
"""
Maze

This is a maze game
"""

playerx = playery = EXITX = EXITY = None
MAZE = WIDTH = HEIGHT = None

PLAYER = '@' # try changing this
WALL = '#'
EMPTY = ' '
EXIT = 'E'
BLOCK = chr(9608) # what the player sees as the wall: '█'

# to find emojis, go to https://unicode.org/emoji/charts/emoji-list.html
# then replace the U+ with 0x and put it in a chr function, like so:
# chr(0x26F3)

WIDTH, HEIGHT, MAZE, playerx, playery, EXITX, EXITY

with open('maze10x10.txt', 'r', encoding='UTF-8') as f:
    MAZE = f.read()
    HEIGHT = len(MAZE.split('\n'))
    WIDTH = len(list(MAZE.split('\n')))

MAZE = MAZE.split('\n')# split the MAZE into rows

for i in range(HEIGHT):
    MAZE[i] = list(MAZE[i])

playerx = 0
playery = 0

EXITX = 0
EXITY = 0

for i in range(HEIGHT):
    for j in range(WIDTH):
        if MAZE[i][j] == 'S':
            playerx = i
            playery = j

        if MAZE[i][j] == 'E':
            EXITX = i
            EXITY = j


def draw_maze():
    """Display the maze:"""
    for x in range(WIDTH):
        for y in range(HEIGHT):
            if (x, y) == (playerx, playery):
                print(PLAYER, end='')
            elif (x, y) == (EXITX, EXITY):
                print('X', end='')
            elif MAZE[x][y] == WALL:
                print(BLOCK, end='')
            else:
                print(MAZE[x][y], end='')
        print()  # Print a newline after printing the row.

while True:
    try:
        draw_maze()

        print('Where would you like to go?')
        print('  W')
        print('A   D')
        print('  S')

        direction = input('> ').lower()

        if direction not in ('w', 'a', 's', 'd', 'quit'):
            print('That is not a valid direction.')
            continue

        if direction == 'a':
            if MAZE[playerx][playery-1] == '#':
                print('That spot is taken, sorry.')
                continue

            playery = playery - 1

        if direction == 'd':
            if MAZE[playerx][playery+1] == '#':
                print('That spot is taken, sorry.')
                continue

            playery = playery + 1

        if direction == 'w':
            if MAZE[playerx-1][playery] == '#':
                print('That spot is taken, sorry.')
                continue

            playerx = playerx - 1

        if direction == 's':
            if MAZE[playerx+1][playery] == '#':
                print('That spot is taken, sorry.')
                continue

            playerx = playerx + 1

        if direction == 'quit':
            break

        if (playerx, playery) == (EXITX, EXITY):
            draw_maze()
            break
    except KeyboardInterrupt:
        break

print('Thank you for playing!')

File: ruby/maze.rb
# set up constants
PLAYER = "@" # how the player is displayed
BLOCK = 9608.chr("UTF-8") # how the walls are shown
WALL = "#" # what a wall is defined by inthe maze
EMPTY = " " # how hallways are defined in the maze
MAZE = [
  "##########".split(""),
  "#S#      #".split(""),
  "#   # # ##".split(""),
  "####   # #".split(""),
  "#        #".split(""),
  "# #  #####".split(""),
  "####     #".split(""),
  "#    #####".split(""),
  "#E#      #".split(""),
  "##########".split("")
]

playerx = -1
playery = -1
exitx = -1
exity = -1

(0..9).each do |x|
  (0..9).each do |y|
    if MAZE[x][y] == "S"
      playerx = x # find the starting point and make the player go there
      playery = y
    end
    if MAZE[x][y] == "E"
      exitx = x # find the exit and set up it's coordinates
      exity = y
    end
  end
end

def draw_maze(px, py)
  (0..9).each do |i|
    (0..9).each do |j|
      if i == px and j == py
        print PLAYER
      elsif MAZE[i][j] == WALL
        print BLOCK
      elsif MAZE[i][j] == EMPTY
        print " "
      else
        print MAZE[i][j]
      end
    end
    puts
  end
  puts
end

while true # main game loop
  draw_maze(playerx, playery)
  
  puts "Where would you like to go?"
  puts "  W"
  puts "A   D"
  puts "  S"
  
  print "> " # prompt the user
  move = gets.chomp.downcase # get user's input, remove whitespace, and lowercase it
  
  if not ["w", "a", "s", "d"].include?(move)
    puts "\033[33mInvalid move." # \033[33m is the ANSI escape code for yellow
    print "\033[0m" # \033[0m is the ANSI escape code for reset
    next
  end
  
  if move == "a"
    if MAZE[playerx][playery - 1] == WALL
      puts "\033[33mYou can't go there, that spot's taken."
      print "\033[0m"
    else
      playery -= 1
    end
  elsif move == "d"
    if MAZE[playerx][playery + 1] == WALL
      puts "\033[33mYou can't go there, that spot's taken."
      print "\033[0m"
    else
      playery += 1
    end
  elsif move == "w"
    if MAZE[playerx - 1][playery] == WALL
      puts "\033[33mYou can't go there, that spot's taken."
      print "\033[0m"
    else
      playerx -= 1
    end
  elsif move == "s"
    if MAZE[playerx + 1][playery] == WALL
      puts "\033[33mYou can't go there, that spot's taken."
      print "\033[0m"
    else
      playerx += 1
    end
  end
  
  if [playerx, playery] == [exitx, exity]
    # if the player is on the exit
    draw_maze(playerx, playery)
    puts "\033[32mYou Win!" # tell them they won
    exit # and exit the program
  end
end
